File: LeetCodePrj/Java/leetcode/easy/page2/ReverseLinkedList.java
package leetcode.easy.page2;

import leetcode.common.ListNode;

/**
 * @author Kyle
 * @create 2018/9/16 11:53
 */
public class ReverseLinkedList {
    public ListNode reverseList(ListNode head) {
        ListNode resultHead = null;

        while (head != null) {
            ListNode tmpNode = head;
            head = head.next;

            tmpNode.next = resultHead;
            resultHead = tmpNode;
        }

        return resultHead;
    }
}

File: LeetCodePrj/Go/src/easy/page2/reverse_linked_list.go
package page2

type ListNode struct {
	Val  int
	Next *ListNode
}

func reverseList(head *ListNode) *ListNode {
	if head == nil || head.Next == nil {
		return head
	}

	tail := head.Next
	curNode := tail.Next
	head.Next = nil
	for {
		tail.Next = head
		head = tail
		tail = curNode
		if tail != nil {
			curNode = curNode.Next
		} else {
			tail = head
			break
		}
	}

	return tail
}
